In [28]:
from sentence_transformers import SentenceTransformer
import numpy as np, re, statistics

bi = SentenceTransformer("all-MiniLM-L6-v2")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 11602.03it/s]
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [29]:
document = """
Franchise Operations Manual.

1. Overview and Purpose

This Franchise Operations Manual establishes the operating requirements, administrative procedures, financial obligations, maintenance standards, store operating rules, and customer transaction policies applicable to all franchise locations. The manual is intended to provide a consistent operating framework while allowing franchisees to manage the day-to-day activities of their individual stores.

All franchisees are responsible for understanding the requirements in this manual before opening a location. Store owners are also responsible for ensuring that managers and employees receive appropriate training for the responsibilities assigned to them. Where this manual identifies an approval requirement, the required approval must be obtained before the related activity takes place unless an emergency exception is explicitly permitted.

Franchisees should use the operations portal for records that the manual specifically requires to be logged or maintained electronically. Records should be accurate, complete, and available for review by authorized representatives of the franchisor.

The requirements in this manual are organized into several major operational areas: onboarding, fees and royalties, equipment and maintenance, store standards, refunds and returns, staffing, inventory management, marketing, compliance, customer service, reporting, and general administrative procedures.

2. Franchisee Onboarding

New franchisees complete a two-week onboarding program before opening. The onboarding program is designed to ensure that the primary owner understands the systems and procedures required to operate a franchise location.

Week one covers point-of-sale training, inventory systems, and staff scheduling. Point-of-sale training introduces the franchisee to transaction processing, payment handling, refunds, discounts, reporting, and basic troubleshooting. Franchisees should become familiar with the procedures for opening and closing the point-of-sale system and should understand how daily transaction information is recorded.

Inventory training during week one covers receiving stock, recording quantities, identifying discrepancies, monitoring inventory levels, and maintaining accurate inventory records. Franchisees are expected to understand the difference between available inventory, reserved inventory, damaged inventory, and inventory that has been removed from sale.

Staff scheduling training covers creating weekly schedules, assigning employees to appropriate shifts, monitoring staffing coverage, and making schedule changes when necessary. Managers should ensure that scheduled staffing levels are appropriate for expected store activity.

Week two covers marketing, local compliance, and a supervised soft launch. Marketing training introduces the franchisee to approved promotional materials, local marketing activities, campaign procedures, and brand communication standards.

Local compliance training covers requirements that may apply to the specific store location, including local permits, operating requirements, signage rules, and other applicable regulations. Franchisees are responsible for understanding the compliance obligations that apply to their location.

The supervised soft launch provides an opportunity to operate the store with oversight before the full opening. During the soft launch, the franchisee and operating team can identify problems with staffing, point-of-sale procedures, inventory handling, customer service, and store operations.

Onboarding must be completed by the primary owner in person. Delegating the entire onboarding program to another employee or manager does not satisfy the primary-owner attendance requirement.

The onboarding program must be completed before the franchise location opens for normal customer operations. Any unresolved onboarding issue should be communicated to the appropriate regional representative before opening.

3. Onboarding Records and Completion

The franchisee should retain confirmation that the onboarding program has been completed. Training records may include attendance information, training modules completed, system access confirmations, and soft-launch observations.

If a franchisee has difficulty completing a training module, the issue should be raised with the assigned training representative. Additional clarification or training may be provided where appropriate.

The primary owner remains responsible for ensuring that the store team understands the procedures necessary to operate the location. Completion of onboarding does not eliminate the franchisee's ongoing responsibility to train new employees or managers.

When a new manager assumes responsibility for a location, the franchisee should ensure that the manager receives appropriate instruction on store procedures, financial controls, inventory handling, customer service, and compliance requirements.

4. Fees and Royalties

Royalty fees are six percent of gross monthly sales, invoiced on the fifth. Gross monthly sales are used as the basis for calculating the royalty obligation. Franchisees should maintain accurate sales records so that the applicable royalty amount can be determined.

The royalty invoice is issued on the fifth of the applicable billing period. Franchisees are responsible for reviewing the invoice and raising any concerns through the appropriate royalty dispute process.

Marketing contributions are two percent of gross sales, billed monthly. Marketing contributions are separate from royalty fees and should not be treated as a substitute for royalty payments.

Both royalty fees and marketing contributions are calculated using gross sales. Franchisees should ensure that sales reporting is complete and consistent with the records maintained in the point-of-sale system.

Late payments incur a one-point-five percent monthly interest charge. Franchisees should monitor invoice due dates and ensure that payments are processed on time. A late payment may result in an additional financial obligation because of the applicable interest charge.

Royalty disputes must be raised within thirty days of invoice. A franchisee who believes that a royalty invoice contains an error should review the relevant sales records and submit the dispute within the required thirty-day period.

Disputes submitted after the thirty-day period may not qualify for review under the standard royalty dispute process. Franchisees should therefore avoid waiting until the end of the billing cycle to investigate discrepancies.

5. Financial Records and Payment Administration

Franchisees should maintain accurate financial records relating to store operations. Sales records, invoices, payment confirmations, and other supporting documentation should be retained according to applicable recordkeeping requirements.

The point-of-sale system should be treated as the primary operational source for daily sales information unless another approved reporting process has been specified.

Managers should not alter financial records without authorization. If an incorrect transaction is identified, the appropriate correction or refund procedure should be followed rather than manually changing historical sales information.

Financial questions that cannot be resolved at the store level should be escalated through the designated franchise support channel.

Franchisees should regularly compare internal sales records with royalty and marketing invoices. Early identification of discrepancies makes it easier to resolve issues within the applicable dispute period.

6. Equipment and Maintenance

Equipment warranties are valid for twenty-four months from installation. The warranty period begins based on the installation date recorded for the applicable equipment.

Franchisees should maintain installation records and equipment documentation so that warranty eligibility can be established when a repair or replacement is required.

Preventive maintenance is required quarterly and logged in the operations portal. Preventive maintenance should not be postponed simply because equipment appears to be functioning normally.

Quarterly maintenance is intended to reduce unexpected equipment failures, identify wear before it becomes a major problem, and maintain reliable store operations.

Maintenance records should include the equipment involved, the date of maintenance, the work performed, and any relevant observations or follow-up requirements. Where the operations portal provides a designated maintenance record, franchisees should use that record rather than relying exclusively on informal notes.

Emergency repairs are reimbursed if pre-approved by the regional manager. Franchisees should obtain approval before authorizing emergency repair work whenever circumstances permit.

A repair should be considered an emergency only when delaying the repair could materially disrupt store operations, create a safety issue, or cause significant equipment damage. The franchisee should provide the regional manager with enough information to evaluate the repair request.

If emergency circumstances make advance approval impractical, the franchisee should document the circumstances and notify the appropriate regional representative as soon as reasonably possible. Reimbursement is not automatic merely because a repair was urgent.

Replacement parts must be sourced from approved vendors only. Franchisees should not purchase unauthorized replacement components solely because they are cheaper or immediately available.

Using non-approved replacement parts can affect equipment reliability, warranty coverage, safety, and maintenance records. Before purchasing a replacement part, the responsible employee should confirm that the vendor is included in the approved vendor list.

7. Equipment Inspection

Store managers should periodically inspect equipment for visible damage, unusual noise, temperature problems, leaks, electrical issues, or other signs of malfunction.

Problems identified during routine inspection should be documented and escalated according to their severity. Minor issues should be addressed through normal maintenance procedures, while urgent failures should be escalated to the regional manager.

Equipment should be kept clean and operated according to manufacturer instructions and franchise operating procedures. Employees should not attempt repairs that require specialized technical knowledge unless they are authorized and qualified to perform the work.

Where equipment is covered by warranty, the franchisee should verify warranty status before arranging a third-party repair. Unauthorized repair work may create complications when warranty service is requested.

8. Store Standards

Stores must operate a minimum of sixty hours per week. The required operating schedule is intended to provide consistent customer access and maintain a predictable store presence.

Store hours are updated under settings, location, and operating schedule. Authorized managers should verify that the operating schedule shown in the system accurately reflects the approved store hours.

Changes to store hours should be made through the designated settings path rather than through informal communication alone. Incorrect operating hours can result in customers receiving inaccurate information about store availability.

Temporary closures require a request submitted to head office at least seven days ahead. The closure request should be submitted as early as possible when the franchisee knows that the store will not be able to operate according to its normal schedule.

A temporary closure may be required because of maintenance, building issues, staffing problems, local events, or other operational circumstances. Regardless of the reason, the required request should be submitted at least seven days before the planned closure whenever the closure is foreseeable.

Unplanned emergencies should be communicated to head office as soon as possible. The seven-day requirement applies to temporary closures that can reasonably be anticipated.

Signage must match brand guidelines version four or later. Store signage should be reviewed periodically to ensure that it remains consistent with the current approved brand presentation.

Franchisees should not independently redesign official signage, logos, colors, typography, or other branded materials. When replacement signage is required, the franchisee should use an approved design and supplier where applicable.

9. Store Opening and Closing

Managers should establish opening and closing routines that provide consistent operational controls. Opening procedures may include checking the store condition, confirming equipment status, preparing the point-of-sale system, verifying inventory requirements, and confirming staff coverage.

Closing procedures should include completing required financial reconciliation, securing equipment and inventory, checking the store condition, and ensuring that access points are properly secured.

Any discrepancy identified during opening or closing should be documented and escalated when necessary.

Employees should not leave the store unattended during required operating hours unless an approved procedure specifically permits it. Staffing requirements should be considered when managers prepare daily schedules.

10. Staffing and Scheduling

Staff scheduling should be completed in advance whenever possible. Managers should consider expected customer traffic, store operating hours, employee availability, required responsibilities, and applicable employment requirements.

Schedules should provide sufficient coverage during opening, closing, peak customer periods, and other operationally important times.

Last-minute absences should be handled through the store's established staffing process. Managers should avoid making schedule changes that leave the store unable to meet its minimum operating requirements.

Employees should receive appropriate training before being assigned independent responsibility for point-of-sale operations, refunds, inventory adjustments, or other controlled processes.

Managers are responsible for supervising employees and ensuring that store procedures are followed consistently.

11. Inventory Management

Inventory should be received, counted, and recorded according to approved store procedures. Any discrepancy between the expected quantity and the quantity physically received should be investigated.

Damaged products should be identified and handled according to the applicable inventory procedure. Employees should not return damaged products to available inventory without authorization.

Inventory counts should be performed at the required frequency and significant discrepancies should be escalated to the store manager.

Managers should monitor products that are frequently out of stock, approaching expiration where applicable, damaged, or otherwise unavailable for sale.

Inventory records should be consistent with physical inventory. Repeated discrepancies may indicate receiving errors, recording errors, theft, damage, or other operational issues and should therefore be investigated.

12. Marketing and Brand Compliance

Local marketing activities should follow approved brand standards. Franchisees may participate in approved local promotions and campaigns provided that the materials and messaging comply with current brand requirements.

Marketing contributions are separate from local marketing expenses incurred by the franchisee. A franchisee should not assume that payment of the monthly marketing contribution automatically authorizes any independent advertising campaign.

Promotional materials should use current brand assets. Outdated logos, signage, slogans, or other materials should be replaced when updated guidelines are issued.

Store managers should ensure that promotional displays are installed correctly and removed when the promotion ends.

13. Local Compliance

Franchisees are responsible for complying with applicable local requirements affecting store operations. Compliance obligations may vary based on location, building, local authority requirements, and the nature of the store.

The franchisee should maintain copies of relevant permits, licenses, inspections, and approvals when required.

Questions about local requirements should be raised before an operational change is implemented. A franchisee should not assume that approval in one location automatically applies to another location.

Compliance issues that could result in a store closure, financial penalty, customer safety issue, or material operational disruption should be escalated promptly.

14. Customer Service Standards

Customers should receive consistent and professional service. Employees should be familiar with basic transaction procedures, refund rules, return requirements, and escalation procedures.

Customer complaints should be handled respectfully and documented when they involve financial disputes, safety concerns, allegations of misconduct, or other significant issues.

Managers should review recurring customer complaints to identify operational problems. Repeated complaints about store hours, product availability, refunds, or staff behavior may indicate a process that requires correction.

15. Refunds and Returns

Refunds are processed within fourteen days to the original payment method. The original payment method should be used whenever the transaction qualifies for a refund.

Returns without a receipt are issued as store credit only. Employees should not provide a cash refund or return funds to an unverified payment method when the customer cannot provide the required receipt.

Refund approvals above five hundred dollars require regional manager sign-off. Store employees and managers should not bypass this approval threshold by splitting a single refund into multiple transactions.

The approval requirement applies to refunds exceeding five hundred dollars. When a refund is near the threshold, employees should verify the transaction amount before completing the refund.

Fraudulent refund attempts are reported to loss prevention. Employees should remain alert to suspicious refund patterns, altered receipts, repeated refund requests, or other behavior that may indicate an attempt to obtain an unauthorized financial benefit.

Employees should not accuse customers of fraud without following the appropriate escalation procedure. Suspicious transactions should be documented and referred to loss prevention for review.

16. Refund Processing Controls

Before processing a refund, employees should verify the original transaction and confirm that the requested refund is consistent with the applicable return policy.

If the transaction cannot be located, the employee should follow the procedure applicable to returns without a receipt rather than creating an unsupported transaction record.

Large-value refunds require additional oversight. A refund above five hundred dollars must receive regional manager sign-off before it is finalized.

Refund records should contain enough information to allow the transaction to be reviewed later. Managers should periodically review refund activity for unusual patterns.

Repeated refunds involving the same employee, customer, product, payment method, or unusual transaction circumstances may warrant additional review.

17. Fraud Prevention

Fraud prevention is a shared responsibility. Employees should follow established transaction controls even when a customer requests an exception.

Common warning signs may include inconsistent customer explanations, altered receipts, unusual payment behavior, repeated attempts to process refunds, requests to refund to a different payment method, or attempts to bypass approval requirements.

Employees should not attempt to personally investigate suspected fraud beyond their assigned responsibilities. Suspicious activity should be documented and escalated to loss prevention.

Loss prevention may request supporting information, transaction records, employee statements, or other documentation as part of an investigation.

18. Records and Documentation

Operational records should be accurate and maintained in the appropriate system. Records relating to maintenance, inventory, financial transactions, refunds, staffing, and store operations may be reviewed as part of routine oversight.

Employees should not intentionally delete, alter, or conceal operational records.

When an error is discovered, the appropriate correction procedure should be followed. Managers should document significant corrections so that the reason for the change can be understood later.

The operations portal should be used for required maintenance records and other operational information where the system provides a designated workflow.

19. Escalation Procedures

Store managers are expected to resolve routine operational issues within their authority. Issues requiring approval or specialized support should be escalated to the appropriate regional or head-office representative.

Regional manager approval is specifically required for emergency repair reimbursement and refunds above five hundred dollars.

Head office involvement is required for planned temporary closures, with the request submitted at least seven days ahead.

Loss prevention should be contacted when fraudulent refund activity or other suspicious transaction behavior is identified.

Escalation should include relevant facts rather than assumptions. When submitting a request, the franchisee or manager should provide transaction details, dates, amounts, supporting records, and a concise description of the issue when applicable.

20. Operational Reviews

Managers should periodically review store performance and operational compliance. Reviews may include store hours, staffing coverage, inventory accuracy, equipment maintenance, refund activity, customer complaints, and outstanding administrative issues.

A review should identify both immediate issues and recurring patterns. A single isolated error may require correction, while repeated errors may indicate a need for additional employee training or process changes.

Franchisees should address operational issues promptly rather than allowing unresolved problems to accumulate.

21. Temporary Store Closures

A temporary closure should be treated as an exception to the normal operating schedule. Planned closures require a request submitted to head office at least seven days ahead.

The request should identify the planned closure period and the reason for the closure. Where possible, the franchisee should also communicate any operational impact that may require additional coordination.

Examples of foreseeable closure-related situations include scheduled building work, planned equipment replacement, major maintenance, or other events that prevent normal operation.

Unexpected events such as serious equipment failure or safety incidents may require immediate closure. In those cases, the franchisee should contact the appropriate support representative as soon as practical.

22. Store Hours and Operating Schedule

The store must operate a minimum of sixty hours per week. Managers should verify the published operating schedule regularly.

Store hours are updated under settings, location, and operating schedule. Managers responsible for maintaining store information should be familiar with this path.

When operating hours change, the manager should verify that the updated schedule is reflected correctly wherever customers may see store hours.

Temporary changes to store hours should not be treated as permanent changes unless the appropriate operating schedule has been formally updated.

23. Brand Standards and Signage

Signage must match brand guidelines version four or later. Franchisees should ensure that exterior and interior signs remain consistent with approved branding.

If a store has signage from an older brand-guideline version, the franchisee should determine whether replacement or modification is required.

Brand compliance includes more than the store logo. Typography, colors, layout, promotional displays, and other visual elements may also be covered by brand guidelines.

Employees should not create unofficial signs that could be mistaken for approved brand materials.

24. Equipment Warranty and Vendor Requirements

Equipment warranties are valid for twenty-four months from installation. Installation dates should therefore be recorded accurately.

When equipment fails, the franchisee should first determine whether the equipment remains within the twenty-four-month warranty period.

If warranty coverage applies, the franchisee should follow the applicable warranty service procedure before paying for an external repair.

Replacement parts must be sourced from approved vendors only. Approved sourcing requirements apply even when a non-approved vendor appears to offer a faster delivery or lower price.

Documentation relating to replacement parts should be retained with the relevant maintenance records where required.

25. Preventive Maintenance Schedule

Preventive maintenance is required quarterly and logged in the operations portal. Quarterly maintenance should be scheduled so that required work is not missed.

Managers should review maintenance records to confirm that the required maintenance has been completed.

If maintenance is delayed, the reason should be documented and the work should be completed as soon as practical.

Preventive maintenance should not be confused with emergency repairs. Preventive maintenance is planned work intended to reduce failures, while emergency repairs address urgent problems requiring immediate attention.

26. Emergency Repairs

Emergency repairs are reimbursed if pre-approved by the regional manager. The approval should be obtained before repair expenses are incurred whenever reasonably possible.

A repair request should provide enough information for the regional manager to understand the problem and assess the urgency.

Franchisees should retain invoices and supporting documentation for approved emergency repairs.

Payment of an emergency repair does not automatically mean that the repair is reimbursable. The pre-approval requirement remains important.

27. Payment and Invoice Disputes

Royalty disputes must be raised within thirty days of invoice. Franchisees should review invoices promptly after they are issued.

A dispute should identify the specific amount or transaction being questioned and provide supporting information where available.

If no dispute is submitted within the applicable thirty-day period, the franchisee may lose the opportunity to challenge the invoice through the standard dispute process.

The invoice date should therefore be treated as the starting point for calculating the dispute period.

28. Interest on Late Payments

Late payments incur a one-point-five percent monthly interest charge. Franchisees should monitor outstanding balances to avoid unnecessary interest charges.

The interest charge is separate from the underlying royalty or marketing obligation. Paying the original invoice after the due date does not necessarily eliminate the applicable interest amount.

Financial administrators should maintain a calendar or other process for monitoring invoice dates and payment deadlines.

29. Store Management Responsibilities

The store manager is responsible for maintaining daily operational discipline. This includes monitoring staffing, store hours, inventory, equipment, customer service, transaction controls, and required documentation.

Managers should ensure that employees understand the policies that apply to their duties.

When an employee is uncertain about whether an action requires approval, the manager should determine the applicable procedure before allowing the transaction or operational change to proceed.

Managers should escalate issues that exceed their authority rather than making unauthorized exceptions.

30. Final Operating Principles

The central operating principle of this manual is consistency. Franchisees should follow established procedures, maintain accurate records, obtain required approvals, and escalate issues when necessary.

The most important recurring requirements include completing the two-week onboarding program before opening, ensuring that the primary owner attends onboarding in person, paying royalty fees equal to six percent of gross monthly sales, paying marketing contributions equal to two percent of gross sales, and addressing royalty disputes within thirty days of invoice.

Equipment warranties remain valid for twenty-four months from installation, preventive maintenance is required quarterly and must be logged in the operations portal, emergency repairs require regional manager pre-approval for reimbursement, and replacement parts must come from approved vendors.

Stores must operate at least sixty hours per week. Store hours are maintained under settings, location, and operating schedule. Planned temporary closures require a request to head office at least seven days ahead, and signage must follow brand guidelines version four or later.

For customer transactions, refunds are processed within fourteen days to the original payment method. Returns without a receipt result in store credit only. Refund approvals above five hundred dollars require regional manager sign-off, while fraudulent refund attempts are reported to loss prevention.

These requirements form the baseline operational expectations for franchise locations. Franchisees should refer to the relevant section of this manual whenever they are uncertain about a financial obligation, approval requirement, maintenance procedure, store operating rule, or customer transaction.
"""

In [30]:
def fixed(text, size=300):
    return [text[i : i + size] for i in range(0 , len(text), size)]


def fixed_overlap(text, size=300, overlap = 80):
    step = size-overlap
    return [text[i : i+size] for i in range(0, len(text), step )]

def recursive(text, size=300):
    paras = [ p.strip() for p in text.split("\n\n") if p.strip()]
    chunks = [ ]
    for p in paras:
        if len(p) <= size:
            chunks.append(p)
        else:
            sents, cur = re.split(r'(?<=[.!?])\s+',p), ""
            for s in sents:
                if len(cur) + len(s) <= size:
                    cur += " " + s
                else:
                    chunks.append(cur.strip())
                    cur = s
            if cur.strip():
                chunks.append(cur.strip())
    return chunks

def semantic(text, threshold=0.5):
    sents = [s.strip() for s in re.split(r'(?<=[.!?])\s+', text.replace("\n"," ")) if s.strip()]
    embs = bi.encode(sents, normalize_embeddings=True)
    chunks, cur = [], [sents[0]]
    for i in range(1, len(sents)):
        sim = float(embs[i]@embs[i-1])
        if sim < threshold:
            chunks.append(" ".join(cur))
            cur = [sents[i]]
        else:
            cur.append(sents[i])
    if cur: chunks.append(" ".join(cur))
    return chunks


strategies = {
    "fixed": fixed(document),
    "fixed_overlap" : fixed_overlap(document),
    "recursive" : recursive(document),
    "semantic" : semantic(document),
}

for name, chunks in strategies.items():
    print(f"{name:14} {len(chunks)} chunks | avg len {int(np.mean([len(c) for c in chunks]))}")


fixed          100 chunks | avg len 298
fixed_overlap  136 chunks | avg len 299
recursive      184 chunks | avg len 160
semantic       193 chunks | avg len 153


In [31]:
indexes = {name : bi.encode(chunks, normalize_embeddings=True) for name, chunks in strategies.items()}

def retrieve(query, name, k=3):
    emb, chunks = indexes[name], strategies[name]
    q = bi.encode([query], normalize_embeddings=True)[0]
    top = np.argsort(-(emb @ q))[:k]
    return [chunks[i] for i in top]

In [32]:
queries = [
    "how long is onboarding", "what are the royalty fees", "when are royalties invoiced",
    "what is the late payment charge", "how long is the equipment warranty",
    "how often is maintenance required", "minimum store hours per week",
    "how do I request a temporary closure", "how long do refunds take",
    "what happens to returns without a receipt", "who approves large refunds",
    "what percent is the marketing contribution", "how do I update store hours",
    "where must replacement parts come from", "how do I dispute a royalty charge",
]

for q in queries[:5]:   # eyeball a few first
    print(f"\n\n Q: {q}")
    for name in strategies:
        top1 = retrieve(q, name, k=1)[0]
        print(f"\n {name:14} → {top1} -> length : {len(top1)}")



 Q: how long is onboarding

 fixed          → complete a two-week onboarding program before opening. The onboarding program is designed to ensure that the primary owner understands the systems and procedures required to operate a franchise location.

Week one covers point-of-sale training, inventory systems, and staff scheduling. Point-of-sale  -> length : 300

 fixed_overlap  → de completing the two-week onboarding program before opening, ensuring that the primary owner attends onboarding in person, paying royalty fees equal to six percent of gross monthly sales, paying marketing contributions equal to two percent of gross sales, and addressing royalty disputes within thir -> length : 300

 recursive      → New franchisees complete a two-week onboarding program before opening. The onboarding program is designed to ensure that the primary owner understands the systems and procedures required to operate a franchise location. -> length : 219

 semantic       → Franchisee Onboarding  Ne

In [36]:
def top_terms(chunk):
    return set(re.findall(r'\w+', chunk.lower()))

def overlap_vs_reference(query, ref="recursive", k=3):
    ref_terms = set().union(*[top_terms(c) for c in retrieve(query, ref, k)])
    row = {}
    for name in strategies:
        got = set().union(*[top_terms(c) for c in retrieve(query, name, k)])
        row[name] = len(got & ref_terms) / len(got | ref_terms)
    return row

agg = {name : [] for name in strategies}
for q in queries:
    for name, val in overlap_vs_reference(q).items():
        agg[name].append(val)

print(" avg content-overlap vs 'recursive' baseline(1.0 = identical retrieval) : ")
for name in strategies:
    print(f"{name:14} {statistics.mean(agg[name]):.2f}")

 avg content-overlap vs 'recursive' baseline(1.0 = identical retrieval) : 
fixed          0.45
fixed_overlap  0.42
recursive      1.00
semantic       0.68


In [40]:
bi2 = SentenceTransformer("BAAI/bge-small-en-v1.5")

ref_chunks = strategies["recursive"]
idx_a = bi.encode(ref_chunks, normalize_embeddings=True)
idx_b = bi2.encode(ref_chunks, normalize_embeddings=True)

def retrieve_terms(query, model, index, k=3):
    q = model.encode([query], normalize_embeddings=True)[0]
    top =np.argsort(-(index @ q))[:k]
    return set().union(*[top_terms(ref_chunks[i]) for i in top])

model_overlaps = []
for q in queries:
    a = retrieve_terms(q,bi,idx_a)
    b=  retrieve_terms(q, bi2, idx_b)
    model_overlaps.append(len(a & b) / len( a | b))

model_swap = statistics.mean(model_overlaps)
chunk_swap = statistics.mean(agg["fixed"])

print(f"overlap when swapping EMBEDDING MODEL (chunking fixed): {model_swap:.2f}")
print(f"overlap when swapping CHUNKING (model fixed):           {chunk_swap:.2f}")
print(f"\n-> lower overlap = bigger disruption to retrieval")
print(f"-> chunking moved retrieval {'MORE' if chunk_swap < model_swap else 'LESS'} than the embedding model")


Loading weights: 100%|██████████| 199/199 [00:00<00:00, 7388.65it/s]
BertModel LOAD REPORT from: BAAI/bge-small-en-v1.5
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


overlap when swapping EMBEDDING MODEL (chunking fixed): 0.76
overlap when swapping CHUNKING (model fixed):           0.45

-> lower overlap = bigger disruption to retrieval
-> chunking moved retrieval MORE than the embedding model


In [41]:
def retrieve_terms_bge(query, model, index, k=3):
    q = model.encode(["Represent this sentence for searching relevant passages: " + query],
                     normalize_embeddings=True)[0]
    top = np.argsort(-(index @ q))[:k]
    return set().union(*[top_terms(ref_chunks[i]) for i in top])

model_overlaps = []
for q in queries:
    a = retrieve_terms(q, bi, idx_a)          # MiniLM, no prefix
    b = retrieve_terms_bge(q, bi2, idx_b)     # BGE, with prefix
    model_overlaps.append(len(a & b) / len(a | b))
print(f"model swap (BGE prefixed): {statistics.mean(model_overlaps):.2f}")

model swap (BGE prefixed): 0.81


### Lab 9 — Chunking Effects (Observation Note)

#### What this lab was about
Measuring how much the CHUNKING strategy determines retrieval, vs how much the EMBEDDING MODEL does. Chunked one ~4k-word structured doc four ways, embedded each, ran 15 queries, and measured content-overlap (Jaccard on retrieved words) to quantify how differently each setup retrieves.

#### The four strategies (cheap/structure-blind → expensive/meaning-aware)
1. Fixed-size, no overlap — cut every N chars; guillotines sentences, chunks start mid-topic.
2. Fixed-size, with overlap — repeats ~80 chars across boundaries; saves boundary-straddling facts but imports neighbor noise.
3. Recursive/structural — split on paragraph→sentence boundaries; respects structure. Production standard (LangChain's RecursiveCharacterTextSplitter).
4. Semantic — embed sentences, cut where consecutive-sentence similarity drops; topically coherent chunks. Smartest, but can over-fragment (some chunks were 76 chars = precise retrieval but thin context).

#### Measured: how much chunking moves retrieval (overlap vs recursive baseline)
- fixed          0.45   (retrieves ~45% same content as recursive → big divergence)
- fixed_overlap  0.42
- recursive      1.00   (baseline)
- semantic       0.68   (closest to recursive, but still ~1/3 different)
Naive fixed-size is the biggest departure; semantic clusters nearest to recursive.

#### The headline: chunking vs embedding model
- Swap embedding model (MiniLM→BGE), chunking fixed:        0.76 overlap (0.81 with BGE query prefix)
- Swap chunking (fixed→recursive), model fixed:             0.45 overlap
- Fair version (semantic→recursive, good-vs-good):          0.68 overlap
Lower overlap = bigger disruption. Chunking (0.45, or 0.68 fair) disrupted retrieval MORE than swapping the embedding model (0.76–0.81). **Chunking is the higher-leverage decision.**

#### Key correction / methodology lessons
- BGE needs a query prefix ("Represent this sentence for searching relevant passages: ") — adding it raised overlap 0.76→0.81, confirming it calibrates BGE's retrieval. Forgetting it silently degrades results.
- A silently-swallowed kwarg typo (normalize_embeddings misspelled) and a copy-paste bug (indexing both sides with the same model) produced a fake "model matters more" result first. Lesson: sentence-transformers ignores unknown kwargs silently — verify your two conditions are actually different before trusting a comparison.
- "I don't know"-style low overlap can be an artifact; always confirm the pipeline is measuring what you think.

#### Magnitude is document-dependent (nuanced interview answer)
On this clean, well-structured doc, even fixed chunks land in roughly the right section, so chunking has less room to wreck things — yet it STILL out-disrupted the model swap. On messy/poorly-structured corpora the gap would be larger. So: chunking generally dominates, but I'd profile the actual corpus rather than assume.

#### Tie to my own system
Portal Assistant uses one-task-per-file, <600-token chunks — effectively a semantic/structural strategy (each file = one coherent unit). This lab is evidence that was a higher-leverage choice than which embedder I picked. Good instinct, now measured.

#### One-line summary
Chunking strategy moves retrieval more than the embedding model does (0.45 vs 0.81 overlap on my doc) — engineers overinvest in "which embedding model" and underinvest in chunking, which is backwards.